### ISSUE TITLE
Issues with multi support

### ISSUE THREAD

**Issue by joerixaop** [petergoldstein/dalli#321]
There are a number of issues with the way the multi method behaves currently.
- It does not check the return value for quiet operations. This is an issue because quiet operations still return "interesting" results, in the case of setq, addq and replaceq those are the failure cases.
- It does not work correctly with multithreading.

E.g. of failure without threads

```
# memcached on local host only accepts values up to 1MB, this makes
# for a reliable error case, other error cases are of course possible in 
# practice, but not as reproducible
memc = Dalli::Client.new(["localhost"], value_max_bytes: 2*1024*1024)
result = memc.multi do memc.set("test", "a" * (2*1024*1024 - 200)); end
p result # true
result = begin memc.get("test"); rescue => e; result = [:error, e]; end
p result # [:error, #<Dalli::DalliError: Response error 3: Value too large>]
result = memc.get("test")
p result # nil # This is the response of the previous get
result = memc.set("test", "a"*10)
p result # nil # This is the response of the previous get
result = memc.get("test")
p result # true  # this is the response of the previous set
result = memc.get("test")
p result # "aaaaaaaaaa"  # This is the response of the previous get
result = memc.set("testeken", "a" * 10)
p result # "\x04\bI\"\x0Faaaaaaaaaa\x06:\x06ET"
```

As you can see the request/response cycle gets completely out of sync, with no way to recover.

Basically what needs to happen is that when a quiet operation is followed by a non-quiet operation that the response must be checked of the quiet operation first. Also at the end of a multi method a noop operation should be sent, so errors can be thrown (note that it must not stop at the first error, it needs to read all the responses, otherwise the cycle will still be out of sync).

I don't think this can work with the current threading behavior though. If a connection is shared between multiple threads there is no way to make sure that the exception is shown in the right thread.

**Comment by panthomakos** [comment]
Any chance of this issue being worked on or the solution revived?  It appears that we've also run into this issue with multi gets and can also reproduce the cases outlined in #561.

**Comment by rgaufman** [comment]
This is quite a critical bug that is impacting us, can you please release a new version?
